File: multiStructure/frameworks/csharp-playwright/Pages/ProductsPage.cs
using Microsoft.Playwright;

namespace PlaywrightFramework.Pages;

public class ProductPage
{
    private readonly IPage _page;

    public ProductPage(IPage page)
    {
        _page = page;
    }

    private ILocator SortingField => _page.Locator(".product_sort_container").First;
    private ILocator SortingDropdown => _page.Locator(".product_sort_container").First;
    private ILocator Title => _page.Locator(".title").First;
    private ILocator CartContainer => _page.Locator("a[class=\"shopping_cart_link\"]");

    public async Task ClickOnSortingContainerAsync()
    {
        await SortingField.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Visible });
        await SortingField.ClickAsync();
    }

    public async Task SelectSortingOptionAsync(string optionText)
    {
        await SortingDropdown.SelectOptionAsync(optionText);
    }

    public async Task<double> GetPriceAndClickOnAddToCartButtonAsync()
    {
        var productItems = _page.Locator(".inventory_item");
        var prices = await productItems.Locator(".inventory_item_price").AllTextContentsAsync();

        double lowestPrice = double.MaxValue;
        int lowestIndex = 0;

        for (int i = 0; i < prices.Count; i++)
        {
            var numPrice = double.Parse(prices[i].Replace("$", ""));
            if (numPrice < lowestPrice)
            {
                lowestPrice = numPrice;
                lowestIndex = i;
            }
        }

        var addToCartButton = productItems.Nth(lowestIndex).Locator("button");
        await addToCartButton.ClickAsync();

        return lowestPrice;
    }

    public async Task ClickOnCartContainerAsync()
    {
        await CartContainer.ClickAsync();
    }

    public async Task<string> GetTitleTextAsync()
    {
        return await Title.TextContentAsync() ?? string.Empty;
    }
}

File: multiStructure/frameworks/js-playwright/pages/products/productsPage.js
class ProductPage {
    constructor(page) {
      this.page = page;
  
      this.sortingField = page.locator('.product_sort_container');
      this.sortingDropdown = page.locator('.product_sort_container');
      this.price = page.locator('inventory-item-price');
      this.title = page.locator('.title');
      this.cartContainer = page.locator('a[class="shopping_cart_link"]');
    }
  
    async clickOnSortingContainer() {
        await this.sortingField.waitFor({ state: 'visible' });
        await this.sortingField.click();
    }
  
    async selectSortingOption(optionText) {
      await this.sortingDropdown.selectOption({ label: optionText });
    }
 
    async getPriceAndClickOnAddToCartButton() {
      const productItems = this.page.locator('.inventory_item');
      const prices = await productItems.locator('.inventory_item_price').allTextContents();
        let lowestPrice = Infinity;
        let lowestIndex = 0;

        prices.forEach((price, index) => {
            const numPrice = parseFloat(price.replace('$', ''));
            if (numPrice < lowestPrice) {
                lowestPrice = numPrice;
                lowestIndex = index;
            }
        });

        const addToCartButton = productItems.nth(lowestIndex).locator('button');
        await addToCartButton.click();

        return lowestPrice;
    }

    async clickOnCartContainer(){
        await this.cartContainer.click();
    }
    
    async getTitleText() {
      return await this.title.first().textContent();
    }
  }
  
  module.exports = { ProductPage };